### PR TITLE
MAINT: use more conservative integer types for umath linalg

### PR DESCRIPTION
**Pull request by argriffing**
I'm not sure this is useful or even works.

### CODE REVIEW

File: numpy/linalg/umath_linalg.c.src [hunk 1]
```diff
@@ -2446,20 +2458,23 @@ init_@lapack_func@(GESDD_PARAMS_t *params,
...
     size_t work_size;
-    size_t iwork_size = 8*((size_t)min_m_n)*sizeof(fortran_int);
+    size_t iwork_size = 8 * safe_min_m_n * sizeof(fortran_int);
 
     if (!compute_urows_vtcolumns(jobz, m, n, &u_row_count, &vt_column_count))
         goto error;
 
-    u_size = ((size_t)u_row_count)*m*sizeof(@ftyp@);
-    vt_size = n*((size_t)vt_column_count)*sizeof(@ftyp@);
+    u_size = ((size_t)u_row_count) * safe_m * sizeof(@ftyp@);
```
Comment (jaimefrio) [inline numpy/linalg/umath_linalg.c.src]: Why casting here? Wouldn't `safe_u_row_count` and `safe_vt_col_count` be more consistent choices?

File: numpy/linalg/umath_linalg.c.src [hunk 2]
```diff
@@ -2558,20 +2573,23 @@ init_@lapack_func@(GESDD_PARAMS_t *params,
...
     if (!compute_urows_vtcolumns(jobz, m, n, &u_row_count, &vt_column_count))
         goto error;
 
-    a_size = ((size_t)m)*((size_t)n)*sizeof(@ftyp@);
-    s_size = ((size_t)min_m_n)*sizeof(@frealtyp@);
-    u_size = ((size_t)u_row_count)*m*sizeof(@ftyp@);
-    vt_size = n*((size_t)vt_column_count)*sizeof(@ftyp@);
+    a_size = safe_m * safe_n * sizeof(@ftyp@);
+    s_size = safe_min_m_n * sizeof(@frealtyp@);
+    u_size = ((size_t)u_row_count) * safe_m * sizeof(@ftyp@);
```
Comment (jaimefrio) [inline numpy/linalg/umath_linalg.c.src]: Same casting vs. `safe_*` variable question here.

**Comment by jaimefrio** [comment]
Mostly LGTM. It certainly cannot hurt...

**Comment by argriffing** [comment]
The latest commit has address some comments about notation consistency.

There is still room for improvement, but this could be out of scope for this PR.  For example, there are some function calls like `underlying_lapack_function(..., (low_precision_int_type) high_precision_workspace_size, ...);` which could be checked for overflow.  Also I think the malloc failure branch is weird -- it sets the output values to zero with no python warnings or exceptions as far as I can tell.  These questions could be related to https://github.com/numpy/numpy/issues/3217.

**Comment by jaimefrio** [comment]
Is this ready to merge then? Or do we want to figure out what exactly is going on with the segfault in #5898 before putting it in?

**Comment by argriffing** [comment]
@jaimefrio I'm pretty sure the reason for the segfault in https://github.com/numpy/numpy/issues/5898 is that
1) After working around some overflow problems, the numpy allocation code no longer attempts to allocate a negative number of things, so it no longer prematurely returns from the function call.  Therefore it actually reaches the point of calling the underlying LAPACK dgesv function.
2) Some implementations of the underlying LAPACK function (e.g. OpenBLAS, apparently) can deal with input matrices of size `NxN`, even with an interface that uses int32 for `N` and even if `N*N` would overflow 32 bits.  For example OpenBLAS uses `blasint` for its interface (e.g. for passing `N`) which sounds like it could be 32-bit, whereas the function that does the actual indexing uses `BLASLONG` which sounds less likely to overflow.
https://github.com/xianyi/OpenBLAS/blob/develop/interface/lapack/gesv.c
https://github.com/xianyi/OpenBLAS/blob/develop/lapack/getrf/getrf_single.c
But the numpy lapack-lite implementation seems to internally use int32 indices into the flat `NxN` memory which overflows and segfaults.
https://github.com/numpy/numpy/blob/maintenance/1.9.x/numpy/linalg/lapack_lite/f2c.h#L10
https://raw.githubusercontent.com/numpy/numpy/maintenance/1.9.x/numpy/linalg/lapack_lite/dlapack_lite.c

My understanding from https://github.com/numpy/numpy/issues/5898#issuecomment-104315105 is that the blas-lite and lapack-lite included with numpy are already known to cause segfaults for these larger matrices (e.g. using `dot`) because of overflow issues like this.

I'm not 100% sure that this is what's going on with the segfault, but this is my best guess.  Maybe wait for comments or a review from @pv or others before merging?

**Comment by pv** [comment]
LGTM (assuming no typos)

**Comment by jaimefrio** [comment]
In it goes then. Thanks!

**Pull request merged**